frontend/src/components/interaction-manager/content/transaction.tsx

```diff
@@ -71,24 +75,43 @@ const TRANSACTION_DESCRIPTIONS: Record<TransactionKeys, string> = {
 
 export const Transaction = ({
   interaction
-}: InteractionContentProps<RequestTransactionSending>) => {
-  const { service } = useGlobal()
-  const transaction = parseTransaction(interaction.event.data)
+}: InteractionContentProps<RequestTransactionReview>) => {
+  const { service, dispatch } = useGlobal()
+  const transaction = useMemo(
+    () => parseTransactionInput(interaction.event),
+    [interaction]
+  )
   const title = TRANSACTION_TITLES[transaction.type]
   const description = TRANSACTION_DESCRIPTIONS[transaction.type]
 
+  useEffect(() => {
```

Comment:
suggestion (non blocking): Just my 2 cents but I think its good to avoid these types of useEffects if possible. Could this action be dispatched in the callback of the event listener? Events/actions that trigger state changes which in turn mount/unmount things which then also trigger other state changes and actions can get messy and hard to debug. I could well be missing some context here though

Reply:
yeah, this got me thinking, it would be a lot nicer to do it that way. the missing context part, which would make it really really difficult to handle it there, is because the events are streamlined in a non-deterministic way, can potentially come from multiple sources at any given time. i'm trying to ensure the way they are presented to the user, forcing them to deal with one by one, which requires some batching and queuing, so in the actual event listener just collects and indexes the incoming events. 
it was kinda hard to figure out what to do with them after, so i just assign a react component to each event type and try to deal with the happening inside there. not a huge fan of the approach, will try to think what i could do to refactor it, Dex and Elmar also pointed out that it's not quite ideal before.